File: code/Java/observer/weather/CurrentConditionsDisplay.java
package observer.weather;

import observer.display.DisplayElement;
import observer.observer.Observer;

public class CurrentConditionsDisplay implements Observer, DisplayElement {
    private WeatherData weatherData;
    private float temperature;
    private float humidity;
    private boolean isObserver;

    public CurrentConditionsDisplay(WeatherData weatherData) {
        this.weatherData = weatherData;
        this.weatherData.registerObserver(this);
        this.isObserver = true;
    }

    @Override
    public void display() {
        System.out.println("CurrentConditions temperature:" + temperature + "F, humidity:" + humidity + "%");
    }

    @Override
    public void update() {
        this.temperature = this.weatherData.getTemperature();
        this.humidity = this.weatherData.getHumidity();
        display();
    }

    @Override
    public void registerObserver() {
        if(!this.isObserver) {
            this.weatherData.registerObserver(this);
            this.isObserver = true;
        }
    }

    @Override
    public void removeObserver() {
        if(this.isObserver) {
            this.weatherData.removeObserver(this);
            this.isObserver = false;
        }
    }
}

File: code/C++/observer/currentconditionsdisplay.cpp
#include "currentconditionsdisplay.h"

CurrentConditionsDisplay::CurrentConditionsDisplay(WeatherData * const w)
{
	this->weatherData = w;
	this->weatherData->registerOserver(this);
	this->isObserver = true;
}
CurrentConditionsDisplay::~CurrentConditionsDisplay()
{
	std::cout << "free CurrentConditionsDisplay" << std::endl;
}
void CurrentConditionsDisplay::update()
{
	this->temperature = this->weatherData->getTemperature();
	this->humidity = this->weatherData->getHumidity();
	display();
}
void CurrentConditionsDisplay::display()
{
	std::cout << "CurrentConditions temperature:" << temperature << "F, humidity:" << humidity << "%" << std::endl;
}
void CurrentConditionsDisplay::registerObserver()
{
	if (!this->isObserver) {
		this->weatherData->registerOserver(this);
		this->isObserver = true;
	}
}
void CurrentConditionsDisplay::removeObserver()
{
	if (this->isObserver) {
		this->weatherData->removeOserver(this);
		this->isObserver = false;
	}
}